### PR TITLE
learning resource drawer v2 run comparison table

### PR DESCRIPTION
**Pull request by gumaerc**
### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/5681#issuecomment-2450078182

### Description (What does it do?)
This PR adds a table to the new learning resource drawer comparing the price and format between runs of a course, if they are different.

### Screenshots (if appropriate):
image
image

### How can this be tested?
 - If you have Posthog set up locally, enable the `lr_drawer_v2` flag
 - If you don't have Posthog set up locally, you can force `drawerV2` to be `true` in `LearningResourceDrawer.tsx`
 - Spin up this branch of `mit-learn`
 - Ensure you have sufficient data backpopulated into your local instance, especially Sloan courses
 - Search for courses that have differing run data:
   - Leading the AI-Driven Organization
   - Becoming a More Digitally Savvy Board Member
   - Managing Product Platforms: Delivering Variety and Realizing Synergies
   - Communication and Persuasion in the Digital Age
 - Ensure that the differing runs table shows up on each of these courses and that the style matches the design
 - Click some other courses on the search page and ensure that they do not show the differing runs table

### CODE REVIEW

File: frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx
Comment (jonkafton) [inline frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx:477]: In other places we display exact dollar amounts without decimals. Check `getDisplayPrecision()` in ol-utilities/src/learning-resources/pricing.ts

File: frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx
Comment (jonkafton) [inline frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx:464]: Would this work better as an html table? Thinking in terms of Voiceover / accessibility.

File: frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx
Comment (jonkafton) [inline frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx:489]: Figma has the address on the next line and no colon https://www.figma.com/design/Eux3guSenAFVvNHGi1Y9Wm/MIT-Design-System?node-id=13904-44433&node-type=instance&m=dev

File: frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx
Comment (jonkafton) [inline frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx:517]: Minor suggestion to pull `<DifferingRunsTable>` to a new file and import it to LearningResourceExpandedV2.tsx as it doesn't use anything in InfoSectionV2.tsx.

**Review by jonkafton — CHANGES_REQUESTED** [review]
Looks good. Veery small changes are to remove decimals from exact dollar amounts and a new line for the location display.

**Comment by jonkafton** [comment]
It's from a previous PR, but is the title section intended to be fixed/sticky? @steven-hatch @mbilalmughal 

Thought it looked a bit odd - could just be me and not an issue on touchscreen. 

image

File: frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx
Comment (ChristopherChudzicki) [reply frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx]: I just looked into this a little bit. My impression is that real HTML tables are nice because screenreaders read row and column headers pretty logically and support arrow key navigation more naturally. (Example: with VoiceOver, VO = voiceover modifier combo... VO + down = navigate down, and VO reads header for new row but not for column, since it's the same).

But with the current layout of this table, it's hard to actually implement as a real HTML table:

Screenshot 2024-11-07 at 1 19 20 PM

We can make it look right by using colspan, but then "Location" loses its association with its row, which defeats the a lot of the purpose of using a real table.

I could be wrong, but I don't see a good way to make this a real table.

Comment (ChristopherChudzicki) [reply frontends/ol-components/src/components/LearningResourceExpanded/InfoSectionV2.tsx]: Also, the current navigation, at least in voiceover, seems pretty fine to me.

VO reads
"November 11, 1250 dollars, in person, location: Cambridge, MA, February 10, ..."

which seems ok.

**Comment by gumaerc** [comment]
> It's from a previous PR, but is the title section intended to be fixed/sticky? @steven-hatch @mbilalmughal
> 
> Thought it looked a bit odd - could just be me and not an issue on touchscreen.
> image

The header is meant to be sticky, but what seems to be happening here is that the image is having z-index precedence over it.

**Comment by gumaerc** [comment]
@jonkafton Thanks for your review. This is ready for another look. @ChristopherChudzicki and I chatted about this some more and we think it's best to not change this to an actual `table` element, at least not for now. Given how screen readers seem to respond to `colspan`, I think we should wait for an accessibility review. If they recommend that we change it, we can take another look at the design with @steven-hatch and @mbilalmughal.

File: frontends/ol-components/src/components/LearningResourceExpanded/DifferingRunsTable.tsx
Comment (jonkafton) [inline frontends/ol-components/src/components/LearningResourceExpanded/DifferingRunsTable.tsx:123]: Is `run.delivery` always an array here as there are conditionals on it above?

File: frontends/ol-components/src/components/LearningResourceExpanded/LearningResourceExpandedV2.tsx
Comment (jonkafton) [inline frontends/ol-components/src/components/LearningResourceExpanded/LearningResourceExpandedV2.tsx:91]: Yay

**Review by jonkafton — APPROVED** [review]
Looking good. Approving with potential change needed for optional chaining on https://github.com/mitodl/mit-learn/pull/1782/files#diff-612b73fc63762015de6077e1ac1b79ec8489f6adf644241585eba9bcfe39adc3R123

File: frontends/ol-utilities/src/learning-resources/pricing.ts
Comment (jonkafton) [inline frontends/ol-utilities/src/learning-resources/pricing.ts:47]: Duplicate?

File: frontends/ol-components/src/components/LearningResourceExpanded/DifferingRunsTable.tsx
Comment (gumaerc) [reply frontends/ol-components/src/components/LearningResourceExpanded/DifferingRunsTable.tsx:123]: Yea, according to the API spec it should always be an array. I've also seen it return an array with a single element:

```ts
delivery: Array<CourseResourceDeliveryInner>
```

File: frontends/ol-components/src/components/LearningResourceExpanded/LearningResourceExpandedV2.tsx
Comment (gumaerc) [reply frontends/ol-components/src/components/LearningResourceExpanded/LearningResourceExpandedV2.tsx:91]: 🙃

File: frontends/ol-utilities/src/learning-resources/pricing.ts
Comment (gumaerc) [reply frontends/ol-utilities/src/learning-resources/pricing.ts:47]: Good catch

**Pull request merged**
